Overhaul Formula/Cask JSON generation

MikeMcQuaid opened the pull request:
Take a slightly hackier but more reliable method of setting up `$HOMEBREW_PREFIX`, `$HOME` and `--appdir`.

- Use constants for placeholders
- Monkeypatch to set `HOMEBREW_PREFIX` consistently to placeholder
- Use environment variable to set `Dir.home` consistently to placeholder
- Use `appdir` short-circuit to set `Cask#appdir` consistently to placeholder
- Use `Cask.generating_hash!` to enable "generating mode" with these patches
- Fix `Formula#caveats` from JSON

See https://github.com/Homebrew/formulae.brew.sh/pull/744 for the formulae.brew.sh side.

Fixes #14505
Fixes #14595

Bo98 reviewed (approved):
Fantastic, thanks! Good we managed to get this in for 4.0.0 as I initially thought it might've been too late.

I think this covers the last of the known cask issues.

The slight snag to be aware of is we might be breaking people until their next auto-update if we're generating the JSON with the new placeholder right away (same scenario as #14611).

Bo98 commented on Library/Homebrew/cask/cask.rb:
```diff
@@ -231,11 +270,15 @@ def eql?(other)
     alias == eql?
 
     def to_h
+      Cask.generating_hash!
```
Do we want this for `to_hash_with_variations` as well?

apainintheneck commented on Library/Homebrew/cask/cask.rb:
```diff
@@ -19,10 +19,48 @@ class Cask
 
     attr_accessor :download, :allow_reassignment
 
+    class << self
+      def generating_hash!
+        return if generating_hash?
+
+        # Apply monkeypatches for API generation
+        @old_homebrew_prefix = HOMEBREW_PREFIX
+        @old_home = Dir.home
+        Object.send(:remove_const, :HOMEBREW_PREFIX)
+        Object.const_set(:HOMEBREW_PREFIX, Pathname(HOMEBREW_PREFIX_PLACEHOLDER))
+        ENV["HOME"] = HOME_PLACEHOLDER
+
+        @generating_hash = true
+      end
+
+      def generated_hash!
+        return unless generating_hash?
+
+        # Revert monkeypatches for API generation
+        Object.send(:remove_const, :HOMEBREW_PREFIX)
+        Object.const_set(:HOMEBREW_PREFIX, @old_homebrew_prefix)
+        ENV["HOME"] = @old_home
+
+        @generating_hash = false
+      end
+
+      def generating_hash?
+        @generating_hash ||= false
+        @generating_hash == true
+      end
+    end
```
How will this handle getting called twice? I ask because the `Cask#to_hash_with_versions` calls `Cask#to_h` internally. Maybe just call it in that method after calling `#to_h`?

MikeMcQuaid replied:
@apainintheneck The `generating_hash?` guards mean it'll be fine being called twice.

apainintheneck reviewed (commented):
In general, I think this is a good idea. A few thoughts...

I wonder what the performance implications are for this when calling `brew info --json=v2 --eval-all` though that's not something we need to necessarily figure out now.

It would be nice to see some regression tests here.

And like @Bo98 I'm a bit confused about this only being enabled by default in `Cask#to_h` when we use `Cask#to_hash_with_variations` to generate the API json.

MikeMcQuaid commented:
> It would be nice to see some regression tests here.

Added.

> And like @Bo98 I'm a bit confused about this only being enabled by default in Cask#to_h when we use Cask#to_hash_with_variations to generate the API json.

Removed from both, will put only in https://github.com/Homebrew/formulae.brew.sh/pull/744 instead.

Rylan12 reviewed (commented):
Thanks, @MikeMcQuaid! This looks great, sorry for the review delay!
